File: record/record.cpp
#include <stddef.h>
#include <memory>

#define  NRECORDS  8388608

class Record {
public:
     void set_id(int i) { id = i; }
private:
     int id;       // the id
};

// Excerpt From
// The C++ Programming Language, Fourth Edition
// Bjarne Stroustrup
// This material may be protected by copyright.

int main(int argc, char *argv[])
{
	auto records = std::make_unique<Record[]>(NRECORDS);
	
	for (size_t i = 0; i < NRECORDS; i++) {
		Record r;

		r.set_id(i);
		records[i] = r;
	}

	return 0;
}

File: record/record.java
class Record {

    int id;

    void setId(int i) {
        id = i;
    }
}

/**
 * The RecordApp class implements an application that
 * creates objects.
 */
class RecordApp {
    static Record[] records;

    public static void main(String[] args) {
        try {
            records = new Record[16777216];

            for (int i = 0; i < records.length; i++) {
                Record r = new Record();

                r.setId(i);
                records[i] = r;
            }
        } catch (OutOfMemoryError e) {
            System.err.println("record: out of memory: " + e.getMessage());
            System.exit(1);
        }
    }
}
